### PR TITLE
Fix alignment issue in DatumToBytea

### PR DESCRIPTION
**Pull request by pykello**
This changes DatumBytea to match postgres style & avoid memcpy between values which are not aligned.

### CODE REVIEW

**Comment by jeff-davis** [comment]
* Is this fixing an aliasing issue, a type punning issue, an alignment issue, or some mix of the above?
* To the extent that it's an aliasing issue, why does that matter, considering that postgres is compiled with -fno-strict-aliasing?
* Why is this being triggered by -Wstrict-aliasing?
* Considering that we are compiling for intel (which doesn't care much about alignment) and that we compile with -fno-strict-aliasing, is this fixing a real issue or just cleanup?

File: src/backend/columnar/cstore_metadata_tables.c [hunk 1]
```diff
@@ -1087,7 +1087,11 @@ DatumToBytea(Datum value, Form_pg_attribute attrForm)
 	{
 		if (attrForm->attbyval)
 		{
-			store_att_byval(VARDATA(result), value, attrForm->attlen);
+			Datum tmp;
+			store_att_byval(&tmp, value, attrForm->attlen);
+
+			memcpy_s(VARDATA(result), datumLength + VARHDRSZ,
+					 &tmp, attrForm->attlen);
```
Comment (onurctirtir) [inline src/backend/columnar/cstore_metadata_tables.c:1094]: should we do something like ?:
```c
ptr += info[dim].typlen;
```


like in https://github.com/postgres/postgres/blob/ca3b37487be333a1d241dab1bbdd17a211a88f43/src/backend/statistics/mcv.c#L873-L886

**Comment by onurctirtir** [comment]
Still seeing the same warning on https://github.com/citusdata/citus-enterprise/commits/release-10.0-Hadi-fix

```c
/10.1.0.citus~pre.20210217.9062783/src/backend/distributed/../columnar/cstore_metadata_tables.c:1091:4: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
    store_att_byval(&tmp, value, attrForm->attlen);
```

File: src/backend/columnar/cstore_metadata_tables.c (hunk 1 above)
Comment (onurctirtir) [reply src/backend/columnar/cstore_metadata_tables.c:1094]: Also, still seeing the same warning on https://github.com/citusdata/citus-enterprise/commits/release-10.0-Hadi-fix

```c
/10.1.0.citus~pre.20210217.9062783/src/backend/distributed/../columnar/cstore_metadata_tables.c:1091:4: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
    store_att_byval(&tmp, value, attrForm->attlen);
```

Comment (pykello) [reply src/backend/columnar/cstore_metadata_tables.c:1094]: In the postgress code it is unpacking multiple values so it needs to advance the pointer. Here we are unpacking only one value.

**Comment by pykello** [comment]
>     * Is this fixing an aliasing issue, a type punning issue, an alignment issue, or some mix of the above?

My understanding is that the compiler thought `VARDATA(bytea)` might not be aligned properly so expansion of `store_att_byval` to `*(Datum *) (T) = (newdatum);` might not be safe if size of Datum is larger than int32 and members of `varattrib_4b` are not aligned to accommodate this. As you noted, this is not an issue in intel. I'm not sure why gcc-5 gives this warning but we are not getting it in later gcc versions.

> 
>     * To the extent that it's an aliasing issue, why does that matter, considering that postgres is compiled with -fno-strict-aliasing?
> 

Packaging tool ignores flags in pg_config and doesn't compile Citus with -fno-strict-aliasing. I get a similar warning in postgres if we remove -fno-strict-aliasing from its compile flags:

```
In file included from ../../../src/include/access/htup_details.h:20,                                   [73/1952]
                 from mcv.c:19:
mcv.c: In function ‘statext_mcv_serialize’:
../../../src/include/access/tupmacs.h:210:6: warning: dereferencing type-punned pointer will break strict-aliasi
ng rules [-Wstrict-aliasing]
  210 |     *(int16 *) (T) = DatumGetInt16(newdatum); \
      |      ^~~~~~~~~~~~~
mcv.c:882:5: note: in expansion of macro ‘store_att_byval’
  882 |     store_att_byval(&tmp, value, info[dim].typlen);
      |     ^~~~~~~~~~~~~~~
../../../src/include/access/tupmacs.h:213:6: warning: dereferencing type-punned pointer will break strict-aliasi
ng rules [-Wstrict-aliasing]
  213 |     *(int32 *) (T) = DatumGetInt32(newdatum); \
      |      ^~~~~~~~~~~~~
mcv.c:882:5: note: in expansion of macro ‘store_att_byval’
  882 |     store_att_byval(&tmp, value, info[dim].typlen);
```

>     * Why is this being triggered by -Wstrict-aliasing?
> 

Since packaging tool removes the `-fno-strict-aliasing`

>     * Considering that we are compiling for intel (which doesn't care much about alignment) and that we compile with -fno-strict-aliasing, is this fixing a real issue or just cleanup?

Not a real issue as far as I understand. Just a clean-up to use the same pattern used in postgres.

**Review by jeff-davis — APPROVED** [review]
Please note in the commit message that this is cleanup to fix a theoretical alignment issue and match the postgres style, and has nothing to do with aliasing.

**Pull request merged**